src/miniscript/lex.rs

```diff
@@ -54,28 +53,22 @@ pub enum Token {
     Sha256,
     Hash256,
     Num(u32),
-    Hash20([u8; 20]),
-    Hash32([u8; 32]),
-    Pubkey(PublicKey),
+    Hash20(&'s [u8]),
+    Bytes32(&'s [u8]),
+    Bytes33(&'s [u8]),
+    Bytes65(&'s [u8]),
```

Comment:
Do you recall if you tried using `[u8; 65]` here to avoid the `'s` lifetime?

Reply:
That would be inefficient. Because we would create an owned `[u8; 65]`. And then pass it to `Pk::from_slice(&sl)` which would create another copy. This avoids it. 

This was not needed earlier because we created the public key object directly while lexing, but now that is awkward for 32-byte keys and would require special-case handling in decode when we would get a `bitcoin::PublicKey` object while parsing taproot keys. The method `Pk::from_slice` fits it nicely with the generic function

Reply:
Wish we had upper MSRV with support for const generics - in this case we can have a generic over different key lengths and also can avoid use of `Pk::from_slice` and have `Pk::from_bytes` with different byte array length per each type...

Reply:
these don't need to be unsized slices though.

you could do &[u8; 32] instead and not copy.

just polyfill this code for casting a slice to an array.,

```rust
#[stable(feature = "try_from", since = "1.34.0")]
impl<T, const N: usize> TryFrom<&[T]> for [T; N]
where
    T: Copy,
{
    type Error = TryFromSliceError;

    fn try_from(slice: &[T]) -> Result<[T; N], TryFromSliceError> {
        <&Self>::try_from(slice).map(|r| *r)
    }
}

#[stable(feature = "try_from", since = "1.34.0")]
impl<'a, T, const N: usize> TryFrom<&'a [T]> for &'a [T; N] {
    type Error = TryFromSliceError;

    fn try_from(slice: &[T]) -> Result<&[T; N], TryFromSliceError> {
        if slice.len() == N {
            let ptr = slice.as_ptr() as *const [T; N];
            // SAFETY: ok because we just checked that the length fits
            unsafe { Ok(&*ptr) }
        } else {
            Err(TryFromSliceError(()))
        }
    }
}
```